ambry-router/src/main/java/com/github/ambry/router/ReplicateBlobOperation.java

```diff
@@ -0,0 +1,490 @@
+/**
+ * Copyright 2022 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.clustermap.DataNodeId;
+import com.github.ambry.clustermap.ReplicaId;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.commons.Callback;
+import com.github.ambry.commons.ResponseHandler;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.network.Port;
+import com.github.ambry.network.RequestInfo;
+import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.protocol.ReplicateBlobRequest;
+import com.github.ambry.protocol.ReplicateBlobResponse;
+import com.github.ambry.quota.QuotaChargeCallback;
+import com.github.ambry.quota.QuotaException;
+import com.github.ambry.quota.QuotaUtils;
+import com.github.ambry.server.ServerErrorCode;
+import com.github.ambry.utils.Time;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class manages the internal state of a {@code ReplicateBlobOperation} during its life cycle.
+ */
+class ReplicateBlobOperation {
+  private static final Logger logger = LoggerFactory.getLogger(ReplicateBlobOperation.class);
+  //Operation arguments
+  private final RouterConfig routerConfig;
+  private final ResponseHandler responseHandler;
+  private final BlobId blobId;
+  private final String serviceId;
+  private final DataNodeId sourceDataNode;
+  private final FutureResult<Void> futureResult;
+  private final Callback<Void> callback;
+  private final Time time;
+  private final NonBlockingRouterMetrics routerMetrics;
+  private final long submissionTimeMs;
+
+  // The operation tracker that tracks the state of this operation.
+  private final OperationTracker operationTracker;
+  // A map used to find inflight requests using a correlation id.
+  private final Map<Integer, RequestInfo> ReplicateBlobRequestInfos;
+  // The result of this operation to be set into FutureResult.
+  private final Void operationResult = null;
+  private final QuotaChargeCallback quotaChargeCallback;
```

Comment:
I have different opinion on this. ReplicateBlob is an internal operation, it should not subject to quota charge.

Reply:
Removed Quota code.